compiler/src/main/kotlin/se/ansman/kotshi/DefaultValueProviders.kt

```diff
@@ -0,0 +1,91 @@
+package se.ansman.kotshi
+
+import com.squareup.javapoet.ParameterizedTypeName
+import javax.lang.model.type.DeclaredType
+import javax.lang.model.type.TypeVariable
+import javax.lang.model.util.Types
+
+class DefaultValueProviders(private val types: Types) {
+    private val providers: MutableList<DefaultValueProvider> = ArrayList()
+
+    fun register(provider: DefaultValueProvider) {
+        providers.add(provider)
+    }
+
+    fun validate() {
+        providers
+                .asSequence()
+                .groupByTo(LinkedHashMap()) { it.type to it.qualifiers }
```

Comment:
This can simply be `providers.groupBy { it.type to it.qualifiers }`

Reply:
It can't since we need the inner list to be mutable

Reply:
Ah, didn't realize it affected the type of the inner list (but makes sense)